### ISSUE TITLE
Quark doesn't detect that the output of first API is the input of the second API 

### ISSUE THREAD

**Issue by 3aglew0** [quark-engine/quark-engine#159]
Hi all, my friends (@Dil3mm3 and @ciastron) and I are working to implement new quark rules for a university semester project (our supervisor is @cryptax). We were analyzing Brazking malware (hash SHA256 be3d8500df167b9aaf21c5f76df61c466808b8fdf60e4a7da8d6057d476282b6, let us know if you want the sample).

We wrote this rule that gets the list of installed applications and put them in the shared preferences:
```
{
    "crime": "Get installed applications and put the list in shared preferences",
    "permission": [],
    "api": [
        {
            "class": "Landroid/content/pm/PackageManager;",
            "method": "getInstalledApplications",
            "descriptor": "(I)Ljava/util/List;"
        },
        {
            "class": "Landroid/content/SharedPreferences$Editor;",
            "method": "putString",
            "descriptor": "(Ljava/lang/String; Ljava/lang/String;)Landroid/content/SharedPreferences$Editor;"
        }
    ],
    "score": 1,
    "label": [
        "package manager",
        "installed application",
        "shared preferences"
    ]
}
```

In the malware sample cited before, this rule achieves 80%, but I think it should get 100%. I don't know if I'm wrong, for this reason I would like you to check with me.

In brazking malware, the rule is detected here:
```
List<ApplicationInfo> installedApplications = getPackageManager().getInstalledApplications(128);
StringBuilder sb = new StringBuilder();
for (ApplicationInfo applicationInfo : installedApplications) {
    if ((1 & applicationInfo.flags) == 0) {
        sb.append(applicationInfo.packageName);
        sb.append(",");
    }
}
// ...
Configuracoes.setString(this, "apps", sb.toString());
```

Where the method ```setString``` of ```Configuracoes``` is:
```
public static void setString(Context context, String str, String str2) {
        SharedPreferences.Editor edit = context.getSharedPreferences(str, 0).edit();
        edit.putString(str, str2);
        edit.apply();
    }
```

So, from the output of ```getInstalledApplications``` it is created a String Builder object, whose string is passed to ```setString``` and then input to ```putString``` (target of our rule).

If Quark implements tainted analysis, shouldn't it detect that the output of the first method is the input of the second, even if there is a String Builder in between? Or is it normal that it isn't detected?

I am using Quark **v21.3.2** .
Thank you!

**Comment by haeter525** [comment]
Hi @3aglew0 ,

The `StringBuilder` was not detected because Quark implements tainted analysis on registers only.
Unfortunately, the state of an instance is not tarced by Quark for now.
The breadcrumb from the first API breaks at the for-loop.
```
List<ApplicationInfo> installedApplications = getPackageManager().getInstalledApplications(128);
StringBuilder sb = new StringBuilder();
for (ApplicationInfo applicationInfo : installedApplications)
    if ((1 & applicationInfo.flags) == 0) {
        sb.append(applicationInfo.packageName); // <-- Breaks here
        sb.append(",");
    }
}
```
To explain better, the object `applicationInfo` comes from a list `installedApplications` created by the first API. As a result, it is tagged as an accomplice by Quark. However, the value of `sb.toString()` is given by the object `sb`, and the object is instanced by a new statement. Thus, Quark thinks it is innocent.

I hope this is helpful for you! Perhaps Quark needs a state-tracing feature for objects.

**Comment by krnick** [comment]
Quark did perform tainted analysis, even if other additional functions are used between two Android native API, such as the `toString()` you mentioned.

I think it's just because the current Quark's implementation for [Bytecode instructions](https://github.com/quark-engine/quark-engine/blob/master/quark/Evaluator/pyeval.py) is not complete, causing the tainted analysis to fail to achieve 100% confidence.

In this case, if the confidence is 100%, then we will check whether the below two methods are using the same register.

You can use `quark -a some.apk -r rule_dir -o result.json` to get these results quickly.

1. `Landroid/content/pm/PackageManager;->getInstalledApplications(I)Ljava/util/List;`

2. `Lcom/gservice/Configuracoes;->setString(Landroid/content/Context; Ljava/lang/Strin  g; Ljava/lang/String;)V`

After checking the smali code manually, it should be our issue due to the lack of bytecode instructions. We will complete these bytecode instructions to the bytecode simulator soon in the next release version.

**Comment by cryptax** [comment]
Not sure to understand your answers. Am I getting this right: @haeter525 says we are not getting 100% because Quark currently only implements tainted analysis for registers and the string with installed packages does not get in a register, but @krnick  believes this could be "fixed"/"improved"?

As for `quark -a some.apk -r rule_dir -o result.json`, the result follows. You'll notice `register` is empty.

```json
{
    "md5": "860d38d83cbaa1be8bd5fa248e078c05",
    "apk_filename": "LocalizeApp.apk",
    "size_bytes": 3183557,
    "threat_level": "Moderate Risk",
    "total_score": 1,
    "crimes": [
        {
            "crime": "Get installed applications and put the list in shared preferences",
            "score": 1,
            "weight": 0.5,
            "confidence": "80%",
            "permissions": [],
            "native_api": [
                {
                    "class": "Landroid/content/pm/PackageManager;",
                    "method": "getInstalledApplications"
                },
                {
                    "class": "Landroid/content/SharedPreferences$Editor;",
                    "method": "putString"
                }
            ],
            "combination": [
                {
                    "class": "Landroid/content/pm/PackageManager;",
                    "method": "getInstalledApplications",
                    "descriptor": "(I)Ljava/util/List;"
                },
                {
                    "class": "Landroid/content/SharedPreferences$Editor;",
                    "method": "putString",
                    "descriptor": "(Ljava/lang/String; Ljava/lang/String;)Landroid/content/SharedPreferences$Editor;"
                }
            ],
            "sequence": [
                {
                    "Lcom/gservice/MainActivity; onCreate (Landroid/os/Bundle;)V": {
                        "first": [
                            "invoke-virtual",
                            "v5",
                            "v0",
                            "Landroid/content/pm/PackageManager;->getInstalledApplications(I)Ljava/util/List;"
                        ],
                        "first_hex": "6e 20 97 02 05 00",
                        "second": [
                            "invoke-static",
                            "v4",
                            "v0",
                            "v5",
                            "Lcom/gservice/Configuracoes;->setString(Landroid/content/Context; Ljava/lang/String; Ljava/lang/String;)V"
                        ],
                        "second_hex": "71 30 a4 75 04 05"
                    }
                }
            ],
            "register": []
        }
    ]
}
```

**Comment by krnick** [comment]
Hi @cryptax 

Yeah, you are right. I believe this is Quark's issue on tained analysis, we will fix it as soon as possible.

For now, this behavior only achieves 80%, but after I check it by reading the smali code, it should be 100% on confidence, and the register column should show up the mutual parent function.

**Comment by 3aglew0** [comment]
Hi all, we (@cryptax, @Dil3mm3, @ciastron and me) are working on a different malware and we encountered an issue similar to the one reported above. Analyzing moqhao malware (sha256: bbee58b1e0787bb851e7f7a4d0c187a8122d68eb67e5fa464696310398ac005b, you can get it here:  https://mega.nz/file/adswiQxQ#T59B9DHbdug3Ans57YejSU6eR4FcAspMYE4bqBEdWrY  the password is: infected ), we wrote this new rule:
```
{
    "crime": "Get all accounts and select by type",
    "permission": [],
    "api": [
	{
	    "class": "Landroid/accounts/AccountManager;",
            "method": "getAccounts",
            "descriptor": "()[Landroid/accounts/Account;"
        },
        {
	    "class": "Ljava/lang/Object;",
            "method": "equals",
            "descriptor": "(Ljava/lang/Object;)Z"
        }
    ],
    "score": 1,
    "label": ["accounts"]
}
```
The rule is detected in Loader.d.a.c class, in method onClick. The interested part of the code is:
``` 
    Account[] accounts = ((AccountManager) systemService).getAccounts();
    // ...
    while (true) {
        // ...
        account = accounts[i2];
        if (d.l.c.i.a("com.google", account.type)) {
            break;
        }
        i2++;
    }
``` 

    and the method d.l.c.i.a does:
``` 
    public static boolean a(Object obj, Object obj2) {
        return obj == null ? obj2 == null : obj.equals(obj2);
    }
```

To sum up, from the array given in output from the first API, the attribute of one element of the array is passed in input to the second API. I think Quark loses the connection between them and reports a confidence of 80% instead of 100%, as happens also in the issue above.

I am using Quark v21.5.1 .
Thank you!
